data/lib/src/datasource_impl/shared_space_document_datasource_impl.dart

```diff
@@ -0,0 +1,98 @@
+// Linagora © 2009–2020. Contribute to Linshare R&D by subscribing to an Enterprise
+// offer!”. You must also retain the latter notice in all asynchronous messages such as
+// e-mails sent with the Program, (ii) retain all hypertext links between LinShare and
+// http://www.linshare.org, between linagora.com and Linagora, and (iii) refrain from
+// infringing Linagora intellectual property rights over its trademarks and commercial
+// brands. Other Additional Terms apply, see
+// <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf>
+// for more details.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
+// more details.
+// You should have received a copy of the GNU Affero General Public License and its
+// applicable Additional Terms for LinShare along with this program. If not, see
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+//
+
+import 'dart:convert';
+import 'dart:io';
+
+import 'package:dartz/dartz.dart';
+import 'package:data/src/datasource/shared_space_document_datasource.dart';
+import 'package:data/src/network/config/endpoint.dart';
+import 'package:data/src/extensions/uri_extension.dart';
+import 'package:data/src/network/model/query/query_parameter.dart';
+import 'package:data/src/network/model/sharedspacedocument/work_group_document_dto.dart';
+import 'package:data/src/util/constant.dart';
+import 'package:domain/domain.dart';
+import 'package:domain/src/model/authentication/token.dart';
+import 'package:domain/src/model/file_info.dart';
+import 'package:domain/src/model/sharedspace/shared_space_id.dart';
+import 'package:domain/src/model/sharedspacedocument/work_group_node_id.dart';
+import 'package:flutter_uploader/flutter_uploader.dart';
+import 'package:rxdart/rxdart.dart';
+
+class SharedSpaceDocumentDataSourceImpl implements SharedSpaceDocumentDataSource {
+  final FlutterUploader _uploader;
+
+  SharedSpaceDocumentDataSourceImpl(this._uploader);
+
+  @override
+  Future<FileUploadState> uploadSharedSpaceDocument(
+      FileInfo fileInfo,
+      Token token,
+      Uri baseUrl,
+      SharedSpaceId sharedSpaceId,
+      {WorkGroupNodeId parentNodeId}) async {
+    final file = File(fileInfo.filePath + fileInfo.fileName);
+    final queryParameters = parentNodeId == null
+        ? <QueryParameter>[]
+        : [StringQueryParameter('parent', parentNodeId.uuid)];
+    final taskId = await _uploader.enqueue(
+        url: baseUrl.withServicePath(Endpoint.sharedSpaces
+            .withPathParameter('${sharedSpaceId.uuid}${Endpoint.nodes}')
+            .withQueryParameters(queryParameters)),
```

Comment:
[Confirm] I just saw `Dio` api has supported add parameters to path? Do we still need our own extension?

Reply:
yes, that my question too. Please help me on that

Reply:
`Dio` supports query with type `Map<String, dynamic> queryParameters,`. So I think using our own is better because what we have done is to limit the type of value that pass to the map to a specific type with smth like `StringQueryParameter` or `BooleanQueryParameter`, ... WDYT?

Reply:
@imGok please also check my fixup commits

Reply:
we should use our own way in upload (not base on Dio client). And for the cases, which base on Dio client, we should convert `List<QueryParameter>` to `Map<String, dynamic>` to use the support from Dio. Are you ok with it?